### PR TITLE
fs: Handle absolute pathname for --stdin-filename

### LINKED ISSUE #1809: Snapshot path for snapshots created with stdin
## Output of `restic version`
restic 0.9.0 (v0.9.0-1-g20bfed59) compiled with go1.10.2 on linux/amd64

## How did you run restic exactly?
```
$ restic -r /tmp/repo -p /tmp/repo_password init
[...]
$ echo foo | restic -r /tmp/repo -p /tmp/repo_password backup --stdin --stdin-filename foo
[...]
$ restic -r /tmp/repo -p /tmp/repo_password ls latest
repository fd5bd830 opened successfully, password is correct
snapshot 8d9d9ac2 of [/home/jonathan/foo] at 2018-05-24 06:00:55.208260662 +0000 UTC):
/foo
```

## What backend/server/service did you use to store the repository?
local

## Expected behavior
The snapshot path (`snapshot <id> of [<path>]`) for snapshots created with stdin should be sensible and not misleading.

## Actual behavior
The snapshot path for stdin snapshots is `<working dir>/<stdin filename>` which likely points to a filepath that never existed which can cause confusion and does not accurately represent the source of the data. 

## Steps to reproduce the behavior
_see above_

## Do you have any idea what may have caused this?
https://github.com/restic/restic/blob/v0.9.0/internal/restic/snapshot.go#L35

> Abs returns an absolute representation of path. **If the path is not absolute it will be joined with the current working directory to turn it into an absolute path**. The absolute path name for a given file is not guaranteed to be unique. Abs calls Clean on the result.
[golang.org](https://golang.org/pkg/path/filepath/#Abs)

## Do you have an idea how to solve the issue?
- Use `/dev/stdin` as the source path. This should not affect how the file is stored within the snapshot (`/<stdin filename>`)

### LINKED ISSUE #2063: --stdin-filename with path separator gives cryptic error message
Output of `restic version`
--------------------------
```
restic 0.9.3 compiled with go1.11.1 on linux/amd64
```

How did you run restic exactly?
-------------------------------

`RESTIC_REPOSITORY` set to S3 bucket+prefix, and `RESTIC_PASSWORD` set appropriately.

```
# sudo -iu postgres pg_dumpall | restic backup -q --stdin --stdin-filename /root/postgresql-dump.sql
Fatal: unable to save snapshot: Lstat: file does not exist
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
S3

Expected behavior
-----------------
The backup should complete successfully, or at least a useful error message should be provided.  I would not expect Restic to need to stat _anything_ as part of this operation; the repository is in S3 and the file data is provided on standard input.

Actual behavior
---------------
Restic gives a "file does not exist" error message.

Steps to reproduce the behavior
-------------------------------
Attempt to upload a file with `--stdin-filename` containing a path separator.

Do you have any idea what may have caused this?
-----------------------------------------------
It appears that Restic joins the provided filename to the current working directory Restic is run from; when I omit `/root/` then the file _is_ stored in the repository as `/root/postgresql-dump.sql` (cwd for the command is `/root`).

Do you have an idea how to solve the issue?
-------------------------------------------
I would suspect this needs a patch to be fixed.  A workaround is to `cd` before running Restic so that it's running from the directory where you want the backed-up file to appear to be stored.

Did restic help you or made you happy in any way?
-------------------------------------------------
Yes, this has become my go-to backup tool for all of my machines.  My company is using it in production to backup multiple critical servers for quick recovery, while also providing inexpensive and easily-updated offsite backups.

### PR DESCRIPTION
**Pull request by garrmcnu**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Return valid directory info from Lstat() for parent directories of the specified filename. Previously only "/" and "." were valid directories.
Also set directory mode as this is checked by archiver.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Closes #2063
Closes #1809

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by fd0** [comment]
Just checking in (since you haven't checked the last box): is this PR ready for review? Are you done? Or do you need help with anything?

**Comment by garrmcnu** [comment]
Thanks for the check-in, yes this is ready for review. If this approach looks good, I believe it does handle absolute pathnames. The only question I had was should relative directory paths be allowed for `--stdin-filename`?
With this fix, a relative directory will still fail because the snapshot path is always pre-pended with "/" and so will always be absolute. But I think supplying a relative path in this context doesn't make sense, so giving an error for a relative directory would seem correct. But good to get other thoughts on that?

**Comment by rawtaz** [comment]
Why not just expand the relative path name so the corresponding absolute one, and use that? Then it's supported and is up to the user to decide how they want to specify the path, and it's in line with how you specify paths in other parts of restic.

**Comment by fd0** [comment]
> Why not just expand the relative path name so the corresponding absolute one, and use that?

I don't think that's a good idea, then the path in the snapshot depends on the current directory, which may be surprising. I'd just take the path as an absolute path, and be done with it:
 *  `/foo/bar` becomes `/foo/bar`
 *  `foo/bar` becomes `/foo/bar`
 *  `./foo` becomes `/foo`
 *  `../../foo/bar` becomes `/foo/bar`

And so on. That's (IMHO) the least surprising thing we can do.

**Comment by fd0** [comment]
Whoopsi, that was the wrong button :)

File: internal/fs/fs_reader_test.go [hunk 1]
```diff
@@ -317,6 +327,49 @@ func TestFSReader(t *testing.T) {
 			test.f(t, fs)
 		})
 	}
+
+	var dirTests = []struct {
```
Comment (fd0) [inline internal/fs/fs_reader_test.go]: I'd split this out into its own test function, maybe `TestFSReaderDir`?

File: internal/fs/fs_reader_test.go [hunk 2]
```diff
@@ -151,8 +152,8 @@ func verifyDirectoryContentsFI(t testing.TB, fs FS, dir string, want []os.FileIn
 }
 
 func checkFileInfo(t testing.TB, fi os.FileInfo, filename string, modtime time.Time, mode os.FileMode, isdir bool) {
-	if fi.IsDir() {
-		t.Errorf("IsDir returned true, want false")
+	if fi.IsDir() != isdir {
```
Comment (fd0) [inline internal/fs/fs_reader_test.go]: Oh, good catch!

File: internal/fs/fs_reader_test.go [hunk 3]
```diff
@@ -151,8 +152,8 @@ func verifyDirectoryContentsFI(t testing.TB, fs FS, dir string, want []os.FileIn
...
+		t.Errorf("IsDir returned %t, want %t", fi.IsDir(), isdir)
```
Comment (fd0) [inline internal/fs/fs_reader_test.go]: We need another test, `os.FileInfo` defines that `Name()` returns the base of the file name, so `path.Base(filename)`:
```suggestion
		t.Errorf("IsDir returned %t, want %t", fi.IsDir(), isdir)
	}
	
	if path.Base(fi.Name()) != fi.Name() {
		t.Errorf("Name() returned is not base: want %q, got %q", path.Base(fi.Name()), fi.Name())
```

File: internal/fs/fs_reader.go [hunk 4]
```diff
@@ -103,17 +103,32 @@ func (fs *Reader) Stat(name string) (os.FileInfo, error) {
 // describes the symbolic link.  Lstat makes no attempt to follow the link.
 // If there is an error, it will be of type *PathError.
 func (fs *Reader) Lstat(name string) (os.FileInfo, error) {
-	switch name {
-	case fs.Name:
-		return fs.fi(), nil
-	case "/", ".":
+	getDirInfo := func(name string) os.FileInfo {
 		fi := fakeFileInfo{
 			name:    name,
```
Comment (fd0) [inline internal/fs/fs_reader.go]: This needs to be `path.Base(name)`:
```suggestion
			name:    path.Base(name),
```

File: internal/fs/fs_reader_test.go (hunk 3 above)
Comment (fd0) [reply internal/fs/fs_reader_test.go]: We also need to check `fi.Name() != path.Base(filename)` below (unfortunately I can't add a suggestion for it).

**Review by fd0 — COMMENTED** [review]
I've added a few comments what we need to change, in general it's a great contribution! In order to handle the relative path issue, I'd suggest passing the path given as `--stdin-filename` through `path.Clean(path.Join("/", opts.StdinFilename))` in `cmd_backup.go:526`. This will do the right thing.

Thanks!

EDIT: I pressed the wrong button (again), this was meant to be a change request.

**Comment by garrmcnu** [comment]
Thanks for the feedback, that sound good, will make those changes.

On the relative path, yes treating it as an absolute path looks like a clean solution and that works, thanks.
One minor point on the suggested change, I believe path.Join() already calls Clean() so don't think it needs to be called on the joined path.

File: internal/fs/fs_reader_test.go (hunk 3 above)
Comment (garrmcnu) [reply internal/fs/fs_reader_test.go]: Ah yes, I completely missed that, thanks!

File: internal/fs/fs_reader_test.go (hunk 1 above)
Comment (garrmcnu) [reply internal/fs/fs_reader_test.go]: yes, that's clearer.

**Comment by garrmcnu** [comment]
Thanks for all the help on this.

And of course, thanks for restic, great backup solution!

**Pull request merged**